zebra/zserv.c

```diff
@@ -1070,19 +1077,26 @@ static void zebra_show_client_brief(struct vty *vty, struct zserv *client)
 		client->v6_route_del_cnt);
 }
 
-struct zserv *zserv_find_client(uint8_t proto, unsigned short instance)
+struct zserv *zserv_find_client_session(uint8_t proto, unsigned short instance,
+					uint32_t session_id)
 {
 	struct listnode *node, *nnode;
 	struct zserv *client;
 
 	for (ALL_LIST_ELEMENTS(zrouter.client_list, node, nnode, client)) {
-		if (client->proto == proto && client->instance == instance)
+		if (client->proto == proto && client->instance == instance &&
+		    client->session_id == session_id)
 			return client;
 	}
 
 	return NULL;
 }
 
+struct zserv *zserv_find_client(uint8_t proto, unsigned short instance)
+{
+	return zserv_find_client_session(proto, instance, 0);
+}
+
 /* This command is for debugging purpose. */
```

Comment:
Be carefull. This imply that the default session_id is equal to 0. I think it will be safer to keep the original code looping only on proto and instance to find the zclient session

Reply:
yes, the default is zero - I think it's necessary to work this way otherwise ordering in the list of client structs will cause conflicts, again: if we don't test the entire tuple, we'll just stop at the first match we find on proto+instance, and that may not be the 'zero' struct. I really want this process to be unambiguous - so the entire tuple is always used in comparison/search - otherwise we'll keep encountering these problems again and again.